README.md
# Site Info Center LIGHT
**Getting information about sites CMS and PHP versions with one click.**

_NOTE: A version of this text in german language, alongside with some screenshots, can be found on my blog: [ https://www.vektorkneter.de/site-information-center-light]( https://www.vektorkneter.de/site-information-center-light)_

If you are a web design / development agency or a freelancer you might be familiar with this scenario: 

You have a bunch of sites with a couple of different content management systems and different server configurations under your care. Maybe hosted on different shared servers, managed by your client himself. And you never know exactly which CMS version and wich PHP version is currently running on each site without having to maintain an internal list. Sounds familiar?

**Site Info Center (SIC) fetches all that information for you with just one click. You only have to configure your sites once in SIC and every time you click SIC's refresh button you get the current version of the CMS and PHP versions in use.**

## Requirements
**SIC** and the **satellite** script (more details in short) are written in PHP.  SIC, the user interface or frontend, uses **CURL** to communicate with the satellite, so your (local) server running SIC must have CURL installed - which is the case in most environments, especially if you local server is driven by XAMPP or MAMP.

**IMPORTANT NOTE: The current version of SIC is intended to be used on a PHP capable server in your local (!) network - your local development server for example - because it has no login / protection / user management that prevents foreign users from seeing sensible sites information. Hence the addition “LIGHT” in the project name. You may use SIC on a remote server and protect it with HTTP BASIC AUTH - but I would not recommend that.**

## How  does it work
**The system is made of two parts:**

* **SIC** The part to be placed onto you local development server where you can reach it with you browser. It provides the user interface and the functionality for fetching information from the sites satellites.

* **Satellite** This is a small PHP script you have to place in the root folder of the sites to be monitored. The satellite answers the request of the SIC with information about the CMS and PHP versions currently in use. The satellite in this project comes with a handful functions for getting version info from CMS I was using or I am still using but can be extended with further functions for the CMS you use. More about this later.

In order to prevent the **satellite** to answer all requests and blasting informations into the wild, we are using a **shared secret** which has to be configured with the site in **SIC** and also has to be placed in the **satellite** script.

If you hit the refresh button on the **SIC** user interface, SIC will call the **satellite**, telling him wich CMS it should search for version information (wich function the satellite should run for getting the CMS version) and providing the shared secret. After the satellite has answered, the received information are displayed in the SIC and also stored in a CSV file in `/history` folder. **Yes, CSV**. There is no need for a database and you could import the CSV files into a spreadsheet tool if you want. SIC also provides a button for **bulk updating** all configured sites and displaying the **version history** of each site with one single click.

Since version **1.4** of **SIC** there will be also a `_summary-latest.csv` created in `/history` folder when you click the red refresh-all-button on the top right. This file will contain all results of the latest sites information bulk update. You will see an according notification when the refresh queue is finished with a download button.

## Configuration: SIC
After downloading the project, you will find a file `sites-config.NEW.php` in the `/sic` folder. Just rename it to `sites-config.php` and configure all your sites using the syntax sample provided in the file:

```php
$sites = array( 
    "example.com" => array(                                     // human readable title of the site to monitor
        "url"       => "https://www.example.com/satellite.php", // full URL of the satellite script
        "sys"       => "PROCESSWIRE",                           // system identifier, the satellite has a function for
        "secret"    => "T0tallY5ecret",                         // the shared secret of the site, HAVE TO match the one in the satellite
        "inact"     => false                                    // set to "true" if the site should not longer monitored but you want access to the history
    ),
    "another-site.com" => array(                                     
        "url"       => "https://www.another-site.com/obscured-filename.php", 
        "sys"       => "WORDPRESS",                                  
        "secret"    => "Y0uN3v3RKn0w",                         
        "inact"     => true                                    
    )
);  

```

If you are done, place the `/sic` folder (you may rename it, of course) on your local server where you can reach it now and in the future. Try reaching SIC with your local URL in the browser - if a list of your configured sites is shown, everything is fine.

### Currently supported CMS

| System identifier | CMS / System                           | available since satellite version |
|-------------------|----------------------------------------|-----------------------------------|
| STATIC            | (static site, just return PHP version) | v0.1                              |
| LEPTON            | Lepton CMS, below version 2.4          | v0.1                              |
| WEBSITEBAKER      | WebsiteBaker                           | v0.2                              |
| WORDPRESS         | WordPress                              | v0.3                              |
| WBCE              | WBCE                                   | v0.4                              |
| PROCESSWIRE       | ProcessWire                            | v0.5                              |
| MODX              | MODX Revolution                        | v0.6                              |
| GETSIMPLE         | GetSimple CMS                          | v0.7                              |
| LEPTON24          | Lepton CMS, since version 2.4          | v0.8                              |
| SHOPWARE          | Shopware, since version 5              | v0.9                              |
| PAGEKIT           | Pagekit, since version 1               | v0.9                              |
| BLACKCAT          | BlackCat CMS, 1.x series               | v0.10                             |
| JOOMLA            | Joomla! CMS, tested with  3.6 & 3.8.2  | v0.11                             |
| CONCRETE5         | Concrete5 CMS, tested w. 8.1.0 & 8.2.1 | v0.12                             |
| JOOMLA15          | legacy Joomla! CMS v1.5                | v0.13                             |
| NEXTCLOUD         | NextCloud                              | v0.15                             |
| PIWIK             | PIWIK / Matomo, tested with 3.9.1      | v0.16                             |

## Configuration: Satellite 
Place a copy of the `satellite.php` (to be found in folder `/satellite` in the download) in the root directory of all your configured sites via FTP.  Update the `$sat_secret` in the satellite to the one you configured for the corresponding site in SIC (don’t use the same secret across all your sites!) and make sure the satellite has a function for your CMS (if not, read the section “Add further CMS functions to satellite”).  You are done.

Try hitting the refresh button next to a site in SIC or the _Refresh All_ button on top right and check if the SIC gets information from the satellite(s).

## Add further CMS functions to satellite
The satellite script comes with a handful functions for getting version info from CMS I was using or I am still using  - at the time of writing this is MODX Revolution, ProcessWire, WordPress and some small, mostly only known in Germany, CMS like WebsitBaker, LEPTON and WBCE (formerly WebsiteBaker Community Edition, now Way Better Content Editing). But you may extend the satellite for the CMS you use and you could also remove functions for CMS you don’t.

The satellite script is quite easy to understand: After checking that the shared secret provided from SIC matches that one set in the satellite, a simple `switch` functions checks the `sys` string provided from SIC an determines which function to be run.

So if you want to add your CMS, just write a new function, deliberate a `sys` string (system identifier) and add both to the `switch` function of the satellite. Afterward you can use your new imagined `sys` string when configuring sites in `config-sites.php` in SIC and the satellite will run your new function.

## Upgrading Site Info Center LIGHT
1. Download the newest version from GitHub
2. Remove file `sites-config.NEW.php` and folder `/history` from the **just downloaded** `sic`-folder (you really don't want to overwrite that file and folder on your local server because they contain your configuration and the version history)
3. Copy the remaining files and folder to the SIC folder on your local server, overwrite old files and folders

This is all. But a previous backup is **always** a great idea.



sic/CHANGELOG.md
# Changelog Site Info Center LIGHT

## v1.9
**12.06.2019**
* fixed an issue where js RefreshComplete() and resets does not fire when a request fails (because Promise.all is not fullfilled)
---

## v1.8
**22.03.2019**
* removed AjaxQ plugin wich was used so far for _sequential_ requests
* now using _simulatinous_ requests to the satellites, making the refresh much faster
* clearing filter, search and sort order of datatables on refresh-all action only if necessary
---

## v1.7
**18.03.2019**
* added DataTables.js for searching and sorting the table of active sites
* added ability to only update all filtered / searched sites
---

## v1.6.2
**21.12.2018**
* some improvements for more visible sites in viewport (font size, icons, padding)
---

## v1.6.1
**03.12.2018**
* some changes in filter markup to prevent js errors if no site is configured yet
---

## v1.6
**30.11.2018**
* added option to filter the active sites table by system
---

## v1.5
**30.11.2018**
* SIC now loads the data from the summary file (`/history/_summary-latest.csv`), which is created on each "refresh all" run. So the table of all active sites shows the latest results when the UI is loaded the first time.
---

## v1.4.1
**05.07.2018**
* added a button to the top nav (next to refresh-all) for download latest summary CSV
---

## v1.4
**04.07.2018**
* added a function for creating a summary CSV file if the bulk refresh (refresh-all button) was started
* fixed output of date and time to correctly adding a leading zero if needed (23:7:5 becomes 23:07:05)
---

## v1.3
**28.11.2017**
* added a badge that shows number of active / inactive sites
* moved changelog from `index.php` in seperate file `CHANGELOG.md` (you are currently reading)
---

## v1.2
**20.11.2017**
* added check for sites-config.php and messege how to create
* made tables responsove with .uk-overflow-auto
---

## v1.1
**19.11.2017**
* introduced progressbar for ajax queue
---
 
## v1.0
**18.11.2017**
* changed UI framework from UIkit 2 to UIkit 3.0.0 (beta)
* modified documents and functions in order to generate UIkit 3 syntax
* removed MagnificPopUp for displaying history, now using UIkit 3 lightbox
---

## v0.6
**15.11.2017**
* removed unused UIkit scripts from /js/components
* removed unused, non-minified UIKit stylesheet from /css
* some typo fixes
* added some setup notes in sites-config.php
---

## v0.5
**14.03.2017**
* added SIC LIGHT version number as PHP variable ($siclight_version)
* some CSS improvements

sic/index.php
<?php
/* Site Info Center LIGHT (SIC LIGHT)
 * author: André Herdling (andreherdling.de)
 *
 * see CHANGELOG.md for changes history
 * */



ini_set('display_errors', 'On');
error_reporting(E_ALL);


$siclight_version = "1.9";

?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>Site Info Center Light</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" href="css/uikit.min.css" />
    <link rel="stylesheet" href="css/sic_light.css" />
</head>
<body>

    <div class="uk-container">

        <h1>Site Info Center LIGHT <small><?=$siclight_version?></small></h1>

        <?php
            if(!file_exists('sites-config.php')){
                echo "
                <div uk-alert class='uk-alert-danger'>
                    <strong>sites-config.php not found!</strong> In order to create this file, just rename the sites-config.NEW.php
                </div>\n";
            } else {
                require_once 'sites-config.php';
                require_once 'includes/functions.php';

                // get active sites table (HTML) and count
                $ActiveSites = ActiveSitesTable($sites);

                // check if a summary file exists in order
                // to show / hide the download-summary-button
                if(file_exists('history/_summary-latest.csv')){
                    $dl_sum_class = 'show';
                } else {
                    $dl_sum_class = 'hide';
                }

                // refresh all / refresh selected button markup
                if($ActiveSites['count']>0){
                    $refresh_all_btn = "<button class='refresh-all uk-button uk-button-danger' type='button' uk-tooltip title='Refresh all active sites'><span uk-icon='icon: refresh'></span> ALL</button>\n";
                    $refresh_selected_btn = "<button class='refresh-selected uk-button uk-button-primary' type='button' uk-tooltip title='Refresh filtered sites'><span uk-icon='icon: refresh'></span> FILTERED</button>\n";
                } else {
                    $refresh_all_btn = "";
                    $refresh_selected_btn = "";
                }


                echo "
                <div class='uk-card uk-card-default'>
                    <div class='uk-card-header'>
                        <div uk-grid class='uk-child-width-expand'>
                            <div>
                                <h2 class='uk-card-title'>Active Sites <span class='uk-badge'>".$ActiveSites['count']."</span></h2>
                            </div>
                            <div class='header-left'>
                                {$refresh_selected_btn}
                                {$refresh_all_btn}
                                <a href='history/_summary-latest.csv' target='blank' class='uk-button uk-button-default download-summary {$dl_sum_class}' uk-tooltip title='Download latest summary CSV'><span uk-icon='icon: download'></span> CSV</a>
                            </div>
                        </div>
                    </div>
                    <div class='uk-card-body'>
                        ".$ActiveSites['table']."
                    </div>
                </div>\n";
                
                echo InactiveSites($sites);   
            }
        ?>

    </div>
 
    <div class="licenses">
        SIC LIGHT <?php echo $siclight_version; ?> by <a href="https://www.andreherdling.de">André Herdling</a> | <a href="licenses.txt">Licenses &amp; used software</a>
    </div>

    <div id="progress">
        <progress id="progressbar" class="uk-progress" value="0" max="100"></progress>
    </div>

    <script src="js/jquery-3.1.1.min.js"></script>
    <script src="js/uikit.min.js"></script>
    <script src="js/uikit-icons.min.js"></script>
    <script src="js/datatables.min.js"></script>
    <script src="js/main.js"></script>
    
</body>
</html>


sic/connector/proxy.php
<?php
ini_set('display_errors', 'On');
error_reporting(E_ALL);
require_once '../sites-config.php';
require_once '../includes/functions.php';

// getting data sent via json POST to this file
$data = json_decode(file_get_contents('php://input'), true);


// if debug_mode is TRUE, we do not test for a valid response and just print
// the response form satellite
$debug_mode = false;


// check if we have data
if(!$data OR $data==''){
    // - NO DATA PROVIDED: Quit Processing
    // setting some data for processing by RespondWithError()
    $data['site_id'] = false;
    $data['site_name'] = false;
    RespondWithError(403,"No information provided");
    die();
} else {
    // - DATA PROVIDED: Ask Satelite
    // getting some basic info from given data
    $site_name = $data['site_name'];
    $site_id = $data['site_id'];

    // getting configuration for this site
    $config = $sites[$site_name];

    // getting url
    $url = $config['url'];

    // creating data to be sent to satellite
    $dataForSat['sys'] = $config['sys'];
    $dataForSat['secret'] = $config['secret'];

    // sending request to satellite
    $sat_response = sendPostRequest($url,$dataForSat);

    // checking HTTP status from response
    if(!$sat_response or $sat_response['statuscode'] == 403){
        RespondWithError(403,"Authorisatzion failed");
        die();
    }
    if(!$sat_response or $sat_response['statuscode'] == 404){
        RespondWithError(404,"Satellite not found");
        die();
    }
    if(!$sat_response or $sat_response['statuscode'] != 200){
        // if the satellite is not found, respond with error
        RespondWithError(400,"Satellite not found on remote host or didn't answered properly");
        die();
    } else {
        // get the answer from satellite in an array
        $sat_response_array = json_decode($sat_response['response'],true);

        // check if we got an array
        if(is_array($sat_response_array) AND isset($sat_response_array['sys_ver']) AND isset($sat_response_array['php_ver']) AND isset($sat_response_array['sat_ver'])){
            // the the answer from satellite and add site_id and site_name
            // which will be used in SIC's javascript
            $siteinfo['site_id'] = $site_id;
            $siteinfo['site_name'] = $site_name;
            $siteinfo['sys'] = $config['sys']; // we need this for CSV history

            // merge the two arrays to get an output
            $output = array_merge($siteinfo, $sat_response_array);

            // respond with json
            http_response_code(200);
            echo json_encode($output);

            // save to results to CSV
            SaveToCSV($output);
        } else {
            if(!$debug_mode){
                RespondWithError(400,"Remote host hasn't answered with valid data");
            } else {
                echo $sat_response['response'];
            }

        }


    }


}

?>

sic/sites-config.NEW.php
<?php
/**
 * Site Info Center LIGHT (SIC LIGHT)
 *
 * SITES CONFIGURATION
 * Setup all your sites you want to monitor witch SIC LIGHT
 * using the following syntax:
 * 
 
    $sites = array( 
        "example.com" => array(                                     // human readable title of the site to monitor
            "url"       => "https://www.example.com/satellite.php", // full URL of the satellite script
            "sys"       => "ProcessWire",                           // system identifier, the satellite has a function for
            "secret"    => "T0tallY5ecret",                         // the shared secret of the site, HAVE TO match the one in the satellite
            "inact"     => false                                    // set to "true" if the site should not longer monitored but you want access to the history
        ),
        "another-site.com" => array(                                     
            "url"       => "https://www.another-site.com/obscured-filename.php", 
            "sys"       => "WORDPRESS",                                  
            "secret"    => "Y0uN3v3RKn0w",                         
            "inact"     => true                                    
        )
    );   
  
 * 
 * SIC LIGHT satellite comes with a few functions for getting version info
 * from a handfull CMS. Feel free to implement own functions for your 
 * CM system in the satellite (!) and add new system identifiers here.
 */





$sites = array(
    
    // set up your sites here
    // you may copy & paste the code from syntax example above as boilerplate

);


sic/connector/history.php
<html>
<head>
    <meta charset="utf-8">
    <title>History</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" href="../css/uikit.min.css" />
    <link rel="stylesheet" href="../css/sic_light.css" />
</head>
<body class="history">
    <div class="uk-container uk-container-center history-container">
        <div class="uk-block uk-block-secondary uk-contrast">
            <div class="uk-container">
            <?php

            if(isset($_GET['name']) and $_GET['name'] !=''){
                $name = urldecode($_GET['name']);

                // specify target file
                $targetFile = '../history/'.$name.'.csv';

                // check if file exisits
                if (file_exists($targetFile)) {
                    // print table
                    echo "<h2>History for <strong>$name</strong></h2>";
                    echo "<table class='uk-table uk-table-divider'>\n\n";
                    $f = fopen($targetFile, "r");
                    while (($line = fgetcsv($f)) !== false) {
                        echo "<tr>";
                        foreach ($line as $cell) {
                            echo "<td>" . htmlspecialchars($cell) . "</td>";
                        }
                        echo "</tr>\n";
                    }
                    fclose($f);
                    echo "\n</table>";
                    echo "<a href='../history/$name.csv' class='uk-button uk-button-primary' type='button' style='margin-top: 20px;' target='blank' rel='noopener'>Download CSV</a>";
                } else {
                    echo "History file does not exisit.";
                }
            } else {
                echo "No data submitted";
            }
            ?>
            </div>
        </div>
    </div>

    <script src="../js/uikit.min.js"></script>
    <script src="../js/uikit-icons.min.js"></script>

</body>
</html>












sic/connector/summarywriter.php
<?php
ini_set('display_errors', 'On');
error_reporting(E_ALL);
require_once '../sites-config.php';
require_once '../includes/functions.php';

// getting data sent via json POST to this file
$data = json_decode(file_get_contents('php://input'), true);



$targetFile = '../history/_summary-latest.csv';

// delete old file
if (file_exists($targetFile)) { 
    unlink($targetFile);
}

// create file and open if not already in place
$fh = fopen($targetFile, 'w');

// create table header in CSV
fputcsv($fh,array('Site','System','Sys Ver','PHP Ver','Sat Ver', 'Date', 'Time'));

// - fallbacks
$sys_ver = "n/a";
$php_ver = "n/a";
$sat_ver = "n/a";
$sys = "n/a";
$date = "n/a";
$time = "n/a";

foreach ($data as $item){
    
    $site_name = $item['site_name'];
    
    // getting configuration for this site
    $config = $sites[$site_name];

    // overwrite fallbacks if there is data
    if(isset($item['sys_ver']) AND $item['sys_ver']!=''){
        $sys_ver = $item['sys_ver'];
    };

    // - PHP version
    if(isset($item['php_ver']) AND $item['php_ver']!=''){
        $php_ver = $item['php_ver'];
    };

    // - satellite version
    if(isset($item['sat_ver']) AND $item['sat_ver']!=''){
        $sat_ver = $item['sat_ver'];
    };

    // - system
    if(isset($item['sys']) AND $item['sys']!=''){
        $sys = $item['sys'];
    } else {
        $sys = $config['sys'];
    };

    // - date
    if(isset($item['date']) AND $item['date']!=''){
        $date = $item['date'];
    } else {
        $date = date('d.m.Y');
    };

    // - time
    if(isset($item['time']) AND $item['time']!=''){
        $time = $item['time'];
    } else {
        $time = date('H:i:s');
    };

    // write data to CSV;
    fputcsv($fh,array($site_name, $sys ,$sys_ver, $php_ver, $sat_ver, $date, $time));

}

// close file handle
fclose($fh);

satellite/satellite.php
<?php
//ini_set('display_errors', 'On');
//error_reporting(E_ALL);
/**
 * SIC Satellite
 *
 * – EN –
 * PLEASE LEAVE THIS FILE IN PLACE
 * This file was placed by your website administrator
 * in order to get information of your website remotely,
 * such as the version of the CMS and the version of PHP.
 * The administrator uses that information for keeping
 * track of technical state of your site.
 * If you have any questions about this file, don't hesitate
 * to ask you administrator named below.
 *
 * – DE –
 * BITTE DIESE DATEI NICHT LÖSCHEN
 * Diese Datei wurde vom Administrator Ihrer Website abeglegt,
 * um aus der Ferne Daten Ihrer Website abzufragen, wie etwa die
 * Version des eingesetzten CMS und PHP-Version
 * Der Administrator nutzt diese Informationen, um den
 * technischen Status Ihrer Website im Blick zu behalten.
 * Falls Sie Fragen zu dieser Datei haben, zögern Sie nicht den
 * unten benannten Administrator Ihrer Site zu befragen.
 *
*/


/*--- SETUP -------------------------------------------------*/
// shared secret, matching the value defined in SIC site_config
$sat_secret = "YOUR_SECRET";



/*--- SATELLITE (no need for changes)------------------------*/
// satellite version: The current version of the satellite
// Will be displayed in your SIC
$siteinfo['sat_ver'] = "0.16";

/**
* see CHANGELOG.md for changes history
*/

// getting php version
$siteinfo['php_ver'] = phpversion();

// check if he got valid data
if(isset($_POST['sys']) AND isset($_POST['secret']) AND $_POST['sys']!='' AND $_POST['secret']!=''){

 // check if secret was correct
 if($sat_secret != $_POST['secret']){
    http_response_code(403);
    echo "Authentification failed.";
 } else {
    // everything seems to be fine, let's proceed ...
    // Determine wich function has to be called
    switch ($_POST['sys']) {
        case "LEPTON24":
            $siteinfo['sys_ver'] = sat_LEPTON24();
            break;
        case "LEPTON":
            $siteinfo['sys_ver'] = sat_LEPTON();
            break;
        case "MODX":
            $siteinfo['sys_ver'] = sat_MODX();
            break;
        case "WORDPRESS":
            $siteinfo['sys_ver'] = sat_WORDPRESS();
            break;
        case "WEBSITEBAKER":
            $siteinfo['sys_ver'] = sat_WB();
            break;
        case "WBCE":
            require_once('config.php'); // if included inside the sat_WBCE() it causes a Fatal Error
            $siteinfo['sys_ver'] = sat_WBCE();
            break;
        case "GETSIMPLE":
            $siteinfo['sys_ver'] = sat_GETSIMPLE();
            break;
        case "PROCESSWIRE":
            $siteinfo['sys_ver'] = sat_PROCESSWIRE();
            break;
        case "STATIC":
            $siteinfo['sys_ver'] = "static";
            break;
        case "SHOPWARE":
            $siteinfo['sys_ver'] = sat_SHOPWARE();
            break;
        case "PAGEKIT":
            $siteinfo['sys_ver'] = sat_PAGEKIT();
            break;
        case "BLACKCAT":
            $siteinfo['sys_ver'] = sat_BLACKCAT();
            break;
        case "JOOMLA":
            $siteinfo['sys_ver'] = sat_JOOMLA();
            break;
        case "JOOMLA15":
            $siteinfo['sys_ver'] = sat_JOOMLA15();
            break;
        case "CONCRETE5":
            $siteinfo['sys_ver'] = sat_CONCRETE5();
            break;
        case "NEXTCLOUD":
            $siteinfo['sys_ver'] = sat_NEXTCLOUD();
            break;
        case "PIWIK":
            $siteinfo['sys_ver'] = sat_PIWIK();
            break;                 
        default:
            http_response_code(400);
            echo "System not valid.";
    }

    // send response
    $response = json_encode($siteinfo);
    echo $response;

 }

} else {
    http_response_code(400);
    echo "No valid data";
};


/**
 * sat_CONCRETE5
 * Gets version of Concrete5, tested with 8.1.0 and 8.2.1
 */
function sat_CONCRETE5(){
    // NOTE: When using Concrete5 Version 5, you HAVE to
    // comment out the include if concrete/bootstrap/configure.php
    // else the satellite will return a rendered output of the page
    $config = include('concrete/bootstrap/configure.php');
    $config = include('concrete/config/concrete.php');
    return $config['version_installed'];
}


/**
 * sat_JOOMLA
 * Gets version of Joomla! CMS, tested with 3.6 and 3.8.2
 */
function sat_JOOMLA(){
    // needs to be defined
    define('_JEXEC', 1);
    if (!defined('_JDEFINES')){
        define('JPATH_BASE', __DIR__);
        require_once JPATH_BASE . '/includes/defines.php';
    }
    require_once JPATH_BASE . '/includes/framework.php';

    // create JVersion object
    $version = new JVersion();

    return $version->getShortVersion();
 }


 /**
 * sat_JOOMLA15
 * Gets version of legacy Joomla! CMS v1.5
 */
 function sat_JOOMLA15(){
    // needs to be defined
    define('_JEXEC', 1);
    if (!defined('_JDEFINES')){
        define('JPATH_BASE', _DIR_);
        require_once JPATH_BASE . '/includes/defines.php';
    }
    require_once JPATH_BASE . '/libraries/joomla/version.php';
    
    // create JVersion object
    $version = new JVersion();
    
    return $version->getShortVersion();
 }


/**
 * sat_BLACKCAT
 * Gets version of BlackCat CMS, 1.x series
 */
function sat_BLACKCAT(){
    require_once('config.php');
    
    include CAT_PATH.'/framework/class.frontend.php';
    $wb = new frontend();

    return CAT_VERSION;
 }


/**
 * sat_PAGEKIT
 * Gets version of Pagekit since version 1
 */
function sat_PAGEKIT(){
    // config file requires a `$path`, setting an empty path variable
    $path = '';
    $config = require('app/system/config.php');
    $version = $config['application']['version'];

    return $version;
}


/**
 * sat_SHOPWARE
 * Gets version of Shopware since version 5
 */
function sat_SHOPWARE(){
    require __DIR__ . '/autoload.php';
    $environment = getenv('SHOPWARE_ENV') ?: getenv('REDIRECT_SHOPWARE_ENV') ?: 'production';
    $kernel = new \Shopware\Kernel($environment, $environment !== 'production');

    return $kernel::VERSION;
}


/**
 * sat_LEPTON24
 * Gets version of LEPTON CMS since version 2.4
 */
function sat_LEPTON24(){
    require_once('config.php');
    require_once(LEPTON_PATH.'/framework/class.frontend.php');
    // Create new frontend object
    $wb = new frontend();

    return  LEPTON_VERSION;
}


/**
 * sat_LEPTON
 * Gets version of LEPTON CMS
 */
function sat_LEPTON(){
    require_once('config.php');

    // connect to db
    $mysqli = new mysqli(DB_HOST, DB_USERNAME, DB_PASSWORD, DB_NAME);
    if ($mysqli->connect_errno) {
        printf("Connect failed: %s\n", $mysqli->connect_error);
        exit();
    }
    // get information
    $sql = "SELECT value FROM ".TABLE_PREFIX."settings WHERE name = 'lepton_version'";

    if($result = $mysqli->query($sql)){
        $row=$result->fetch_assoc();
        return $row['value'];
    } else {
        return "not found";
    };
}


/**
 * sat_WB
 * Gets version of WebsiteBaker CMS
 */
function sat_WB(){
  require_once('config.php');

  // connect to db
  $mysqli = new mysqli(DB_HOST, DB_USERNAME, DB_PASSWORD, DB_NAME);
  if ($mysqli->connect_errno) {
      printf("Connect failed: %s\n", $mysqli->connect_error);
      exit();
  }
  // intinal values
  $wb_version = "not found";
  $wb_sp = "";
  $wb_revision = "";

  // get version information
  $sql = "SELECT * FROM ".TABLE_PREFIX."settings WHERE name = 'wb_version'";
  if($result = $mysqli->query($sql)){
  	$row=$result->fetch_assoc();
    $wb_version = $row['value'];
  };

  // get service pack information
  $sql = "SELECT * FROM ".TABLE_PREFIX."settings WHERE name = 'wb_sp'";
  if($result = $mysqli->query($sql)){
  	$row=$result->fetch_assoc();
    $wb_sp = " ".$row['value'];
  };

  // get revision information
  $sql = "SELECT * FROM ".TABLE_PREFIX."settings WHERE name = 'wb_revision'";
  if($result = $mysqli->query($sql)){
  	$row=$result->fetch_assoc();
    $wb_revision = " Rev".$row['value'];
  };

  // combine version
  $version = $wb_version.$wb_sp.$wb_revision;
  return $version;
}

/**
 * sat_WORDPRESS
 * Gets version of WordPress
 */
function sat_WORDPRESS(){
  require_once('wp-includes/version.php');
  return $wp_version;
}

/**
 * sat_WBCE
 * Gets version of WebsiteBaker Community Edition
 */
function sat_WBCE(){
    // connect to db
    $mysqli = new mysqli(DB_HOST, DB_USERNAME, DB_PASSWORD, DB_NAME);
    if ($mysqli->connect_errno) {
        printf("Connect failed: %s\n", $mysqli->connect_error);
        exit();
    }

    // get information
    $sql = "SELECT * FROM ".TABLE_PREFIX."settings WHERE name = 'wbce_version'";

    if($result = $mysqli->query($sql)){
        $row=$result->fetch_assoc();
        return $row['value'];
    } else {
        return "not found";
    };
}

/**
 * sat_PROCESSWIRE
 * Gets version of ProcessWire
 */
function sat_PROCESSWIRE(){
  // loading the API
  require_once('index.php');
  return $wire->config->version;
}

/**
 * sat_MODX
 * Gets version of MODX Revolution
 */
function sat_MODX(){
  // loading MODX
  require_once 'config.core.php';
  require_once MODX_CORE_PATH.'model/modx/modx.class.php';
  $modx = new modX();
  $modx->initialize('web');
  $modx->getService('error','error.modError', '', '');

  // getting version
  $vers = $modx->getVersionData();
  return $vers['full_version'];
}

/**
 * sat_GETSIMPLE
 * Gets version of GetSimple CMS
 */
function sat_GETSIMPLE(){
    define("IN_GS",true);
    require_once('admin/inc/basic.php');
    require_once('admin/inc/configuration.php');
    return $site_version_no;
}

/**
 * sat_NEXTCLOUD
 * Gets version of NextCloud
 */
function sat_NEXTCLOUD(){
    require_once('version.php');
    return $OC_VersionString;
}

/**
 * sat_PIWIK
 * Gets version of PIWIK / Matomo
 */
function sat_PIWIK(){
    require_once('core/Version.php');
    $version = new Piwik\Version;
    return $version::VERSION;
}